### PR TITLE
Fix parallel creating commit status bug with tests

### PR DESCRIPTION
**Pull request by lunny**
This PR is a follow up of #21469

### CODE REVIEW

File: tests/integration/repo_commits_test.go [hunk 1]
```diff
@@ -51,6 +53,18 @@ func doTestRepoCommitWithStatus(t *testing.T, state string, classes ...string) {
 	// Call API to add status for commit
 	t.Run("CreateStatus", doAPICreateCommitStatus(NewAPITestContext(t, "user2", "repo1"), path.Base(commitURL), api.CommitStatusState(state)))
 
+	runBody := doAPICreateCommitStatus(NewAPITestContext(t, "user2", "repo1"), path.Base(commitURL), api.CommitStatusState(state))
```
Comment (delvh) [inline tests/integration/repo_commits_test.go]:
```suggestion
	// Call API to add status for commit, even in parallel
	runBody := doAPICreateCommitStatus(NewAPITestContext(t, "user2", "repo1"), path.Base(commitURL), api.CommitStatusState(state))
	t.Run("CreateStatus", runBody)
	
```
Should hopefully not change any behavior, unless we depend on the test context being unique…

**Review by delvh — COMMENTED** [review]
The second `test` in the PR title seems redundant.

Also, should we label this PR as `skip-changelog`, or is that implied in `kind/testing`?

**Comment by delvh** [comment]
Also, a comment why we ignore Sqlite would be good.

**Comment by lunny** [comment]
> Is this PR alive or dead?

Please review again.

**Comment by lunny** [comment]
> Is this PR alive or dead?

Please review again.

**Review by wolfogre — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-M work

**Comment by wxiaoguang** [comment]
Some hints:

For postgresql, if the `INSERT INTO`  encounters duplicated key and reports error, the transaction can not continue any more. So this solution is flawed for postgresql (in rare cases). For other databases (mssql, mysql, sqlite3), if the `INSERT INTO` fails, the transaction can still continue, so the  approach could  work.

**Comment by harryzcy** [comment]
Does this PR introduce a `DATA RACE` error in the integration test?

```
==================
WARNING: DATA RACE
Read at 0x000008cb6420 by goroutine 239777:
  code.gitea.io/gitea/tests/integration.getTokenForLoggedInUser()
      /drone/src/tests/integration/integration_test.go:268 +0x54
  code.gitea.io/gitea/tests/integration.NewAPITestContext()
      /drone/src/tests/integration/api_helper_for_declarative_test.go:36 +0x8c
  code.gitea.io/gitea/tests/integration.TestRepoCommitsStatusParallel.func1.1()
      /drone/src/tests/integration/repo_commits_test.go:142 +0x84
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x40

Previous write at 0x000008cb6420 by goroutine 239776:
  code.gitea.io/gitea/tests/integration.getTokenForLoggedInUser()
      /drone/src/tests/integration/integration_test.go:268 +0x70
  code.gitea.io/gitea/tests/integration.NewAPITestContext()
      /drone/src/tests/integration/api_helper_for_declarative_test.go:36 +0x8c
  code.gitea.io/gitea/tests/integration.TestRepoCommitsStatusParallel.func1.1()
      /drone/src/tests/integration/repo_commits_test.go:142 +0x84
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x40

Goroutine 239777 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x564
  code.gitea.io/gitea/tests/integration.TestRepoCommitsStatusParallel.func1()
      /drone/src/tests/integration/repo_commits_test.go:141 +0x138
  code.gitea.io/gitea/tests/integration.TestRepoCommitsStatusParallel.func2()
      /drone/src/tests/integration/repo_commits_test.go:147 +0x54

Goroutine 239776 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x564
  code.gitea.io/gitea/tests/integration.TestRepoCommitsStatusParallel.func1()
      /drone/src/tests/integration/repo_commits_test.go:141 +0x138
  code.gitea.io/gitea/tests/integration.TestRepoCommitsStatusParallel.func2()
      /drone/src/tests/integration/repo_commits_test.go:147 +0x54
==================
--- FAIL: TestRepoCommitsStatusParallel (0.88s)
```

**Comment by lunny** [comment]
> 

I think you are right.
```
 testlogger.go:77: 2022/12/01 06:47:08 ...pi/v1/repo/status.go:65:NewCommitStatus() [E] [63884dec-12] CreateCommitStatus: NewCommitStatus[repo_id: 1, user_id: 2, sha: 65f1bf27bc3bf70f64657658635e66094edbcb4d]: generate commit status index failed: pq: current transaction is aborted, commands ignored until end of transaction block
    testlogger.go:77: 2022/12/01 06:47:08 ...eb/routing/logger.go:98:func1() [I] [63884dec-12] router: completed POST /api/v1/repos/user2/repo1/statuses/65f1bf27bc3bf70f64657658635e66094edbcb4d?token=44f685f4ffd298711136bc5065ee155259e15ba8 for , 500 Internal Server Error in 181.6ms @ repo/status.go:20(repo.NewCommitStatus)
    --- FAIL: TestRepoCommitsStatusParallel/ParallelCreateStatus_2 (0.47s)
        pull_status_test.go:104: 
            	Error Trace:	/drone/src/integration_test.go:336
            	            				/drone/src/integration_test.go:165
            	            				/drone/src/pull_status_test.go:104
            	            				/drone/src/repo_commits_test.go:141
            	Error:      	Not equal: 
            	            	expected: 201
            	            	actual  : 500
            	Test:       	TestRepoCommitsStatusParallel/ParallelCreateStatus_2
            	Messages:   	Request: POST /api/v1/repos/user2/repo1/statuses/65f1bf27bc3bf70f64657658635e66094edbcb4d?token=44f685f4ffd298711136bc5065ee155259e15ba8
        pull_status_test.go:104: Response: {"message":"","url":"http://localhost:3002/api/swagger"}
```

**Comment by lunny** [comment]
> 

How did you test that? I think we have enabled RACE_ENABLED in our CI, but I couldn't find the log

**Comment by harryzcy** [comment]
It's in my PR, https://drone.gitea.io/go-gitea/gitea/63992. I got the issue after merge from `main`.

**Pull request merged**